fake_joint_driver/src/fake_joint_driver.cpp

```diff
@@ -25,103 +24,145 @@ FakeJointDriver::FakeJointDriver(void)
+       it++) {
+    for (auto i = 0; i < joint_names_.size(); i++) {
+      if (joint_names_[i] == it->first) {
         act_dis[i] = it->second;
         cmd_dis[i] = it->second;
+        cmd_dis_old[i] = cmd_dis[i];
       }
     }
   }
 
   // Create joint_state_interface and position_joint_interface
-  for (int i = 0; i< joint_names_.size(); i++)
-  {
+  for (int i = 0; i < joint_names_.size(); i++) {
     ROS_DEBUG_STREAM("joint[" << i << "]:" << joint_names_[i]);
     // Connect and register the joint_state_interface
-    hardware_interface::JointStateHandle state_handle(joint_names_[i], &act_dis[i], &act_vel[i], &act_eff[i]);
+    hardware_interface::JointStateHandle state_handle(
+        joint_names_[i], &act_dis[i], &act_vel[i], &act_eff[i]);
     joint_state_interface.registerHandle(state_handle);
 
     // Connect and register the position_joint_interface
-    hardware_interface::JointHandle pos_handle(joint_state_interface.getHandle(joint_names_[i]), &cmd_dis[i]);
+    hardware_interface::JointHandle pos_handle(
+        joint_state_interface.getHandle(joint_names_[i]), &cmd_dis[i]);
     position_joint_interface.registerHandle(pos_handle);
+
+    // Connect and register the velocity_joint_interface
+    hardware_interface::JointHandle vel_handle(
+        joint_state_interface.getHandle(joint_names_[i]), &cmd_vel[i]);
+    velocity_joint_interface.registerHandle(vel_handle);
   }
   registerInterface(&joint_state_interface);
   registerInterface(&position_joint_interface);
+  registerInterface(&velocity_joint_interface);
 }
 
-FakeJointDriver::~FakeJointDriver()
-{
+FakeJointDriver::~FakeJointDriver() {}
+
+/**
+ * @brief Update function to call all of the update function of motors
+ */
+void FakeJointDriver::update(ros::Duration period) {
+
+  // create a lambda substract functor
+  std::function<double(double, double)> substractor = [](double a, double b) {
+    return std::abs(a - b);
+  };
+
+  // create a position difference vector
+  std::vector<double> pos_diff;
+  pos_diff.resize(cmd_dis.size());
+  std::transform(cmd_dis.begin(), cmd_dis.end(), cmd_dis_old.begin(),
+                 pos_diff.begin(), substractor);
+
+  double pos_diff_sum = 0.0;
+  std::for_each(pos_diff.begin(), pos_diff.end(),
+                [&pos_diff_sum](double a) { return pos_diff_sum += a; });
+
+  // if difference exist between two consecutive position commands use position
+  // interface
+  if (pos_diff_sum != 0.0) {
+    // TODO enforce position and velocity limits
+    // only do loopback
+    act_dis = cmd_dis;
+    std::fill(cmd_vel.begin(), cmd_vel.end(), 0.0);
+  } else {
+    // TODO enforce position, velocity and acceleration limits
+    // if there is NO difference between two consecutive position commands use
+    // velocity interface
+    act_vel = cmd_vel;
+    // create a lambda integration functor
+    std::function<double(double, double)> fcn = [&period](double a, double b) {
```

Comment:
let's name it something besides fcn. Maybe `integrator`?

Reply:
https://github.com/PickNikRobotics/fake_joint/pull/1/commits/a1941184a48cb560eab2f970460d4cb4d587a669